# Proteomic-and-transcriptomic-signatures-of-microglia-derived-extracellular-vesicles-
This Repository hosts code that was used in analyzing data for the paper: Identification of state-specific proteomic and transcriptomic signatures of microglia-derived extracellular vesicles

Identification of state-specific proteomic and transcriptomic signatures of microglia-derived extracellular vesicles 


Juliet V. Santiago1,2, Aditya Natu1,2, Christina Ramelow1,2, Sruti Rayaprolu1,2, Hailian Xiao1,2, Vishnu Kumar1,2, Nicholas T. Seyfried1,2,3, Srikant Rangaraju1,2  

 
1. Department of Neurology, Emory University,201 Dowman Drive Atlanta, Georgia, 30322, United States of America   

2. Center for Neurodegenerative Diseases, Emory University, Atlanta, GA 30322, USA  

3. Department of Biochemistry, Emory University, Atlanta, GA 30322, USA  

 

Corresponding author: Srikant Rangaraju Srikant.rangaraju@emory.edu & srikant.rangaraju@yale.edu

ABSTRACT: Microglia are resident immune cells of the brain that play important roles in mediating inflammatory responses in several neurological diseases via direct and indirect mechanisms. One indirect mechanism may involve extracellular vesicle (EV) release, so that the molecular cargo transported by microglia-derived EVs can have functional effects by facilitating intercellular communication. The molecular composition of microglia-derived EVs, and how microglial activation states impacts EV composition and EV-mediated effects in neuroinflammation, remain poorly understood. We hypothesize that microglia-derived EVs have unique molecular profiles that are determined by microglial activation state. Using size exclusion chromatography to purify EVs from BV2 microglia, combined with proteomic (label-free quantitative mass spectrometry or LFQ-MS) and transcriptomic (mRNA and non-coding RNAseq) methods, we obtained comprehensive molecular profiles of microglia-derived EVs. LFQ-MS identified several classic EV proteins (tetraspanins, ESCRT machinery, and heat shock proteins), in addition to over 200 proteins not previously reported in the literature. Unique mRNA and microRNA signatures of microglia-derived EVs were also identified. After treating BV2 microglia with lipopolysaccharide (LPS), interleukin-10, or transforming growth factor beta, to mimic pro-inflammatory, anti-inflammatory, or homeostatic states, respectively, LFQ-MS and RNAseq revealed novel state-specific proteomic and transcriptomic signatures of microglia-derived EVs. Particularly, LPS treatment had the most profound impact on proteomic and transcriptomic compositions of microglia-derived EVs. Furthermore, we found that EVs derived from LPS-activated microglia were able to induce pro-inflammatory transcriptomic changes in resting responder microglia, confirming the ability of microglia-derived EVs to relay functionally-relevant inflammatory signals. These comprehensive microglia-EV molecular datasets represent important resources for the neuroscience and glial communities, and provide novel insights into the role of microglia-derived EVs in neuroinflammation.  

TERMS: Extracellular vesicles, microglia, RNA sequencing, mass spectrometry, proteomics, transcriptomics, neuroinflammation 

 

 
